### DISCUSSION TITLE
UDM File Format Support

### DISCUSSION THREAD

**Discussion by markfortner** [cdk/cdk#1058] [Q&A]
I was wondering if there were any plans to support the UDM file format in CDK?

**Comment by johnmay** [comment]
No plans.

I would never say never but I’m not convinced on why it would be useful. Any outline on why you think we should support it?

Last time I checked I didn’t think it was that great and made some strange choices, was inefficient and bloated. There are other (competing?) initiatives like the ORD which uses Google protocol buffers.

**Comment by egonw** [comment]
> I was wondering if there were any plans to support the UDM file format in CDK?

@markfortner, can you tell me a bit more? Would you need a reader, a writer, or both? Some formats we have IO for (like XYZ) have minimal information and we can more easily write them than read them, for example.

**Comment by markfortner** [comment]
We’d looking at adding support for reactions to our chemistry module.  It currently supports requesting compounds and libraries, but we’d like to support library design and small compound set design (typically favoured by medicinal chemists). We also want to support requests where scaffolds are used. 

Here are our current thoughts/questions:

1. Is there a general model for reactions in CDK? If there is, and it is loosely coupled, then support for a serializer pattern would be straight forward, making it possible to support different file formats.
2. We want to be able to import multi-step reactions, and want to be able to visualize those reactions.
3. We have reactions that are more “formulaic” in nature, where the chemists want to specify a range of input reactants and have the ability to enumerate the products. The UDM examples that we saw, were more in the nature of “1+2=3” as opposed to “A+B=C” and it’s this latter formulaic representation that we’re interested in.
4. We want to be able to support both on-DNA and off-DNA reactions. Previously, we’ve ignored the DNA barcode when representing on-DNA building blocks, so this might not be an issue, but it would be good to know how others are representing them.
5. We looked at UDM because there’s a library of reactions that users could import into our system. I’m not sure if there are other reaction libraries that are available, and their formats and licensing constraints. Given it’s support by the Pistoia Alliance, it seems like a format that might have some traction in industry and might be used across different types of commercial software. But we’re interested in learning about the formats that are currently used and the relative popularity of each one.
6. It would be nice if there was support for stereochemistry as well. Even though many catalogues don’t include stereochemistry information.

**Reply by johnmay** [reply]
The underlying data structures are essentially there for multi-step reactions (IReactionSet). Note as of recently you can also load these via SMILES, see example on CDK Depict:

![image](https://github.com/cdk/cdk/assets/983232/4ddb64a5-40bc-4ee5-bdbc-538107df5206)

My view is most of the UDM guff is probably better fitted in a downstream application rather than a format with first-class support in the CDK. We are happy to help with some required representation.
